greenlet.js

```diff
@@ -1,25 +1,55 @@
+/**
+ * Checks whether an object implements a Transferable interface
+ * @param {*} obj
+ * obj to check for Transferable interface
+ */
+function isTransferable(obj) {
+	// https://developer.mozilla.org/en-US/docs/Web/API/Transferable
+	return (
+		obj instanceof ArrayBuffer ||
+		obj instanceof MessagePort ||
+		obj instanceof ImageBitmap
+	);
+}
+
 /** Move an async function into its own thread.
  *  @param {Function} asyncFunction  An (async) function to run in a Worker.
  *  @public
  */
 export default function greenlet(asyncFunction) {
 	// Create an "inline" worker (1:1 at definition time)
 	let worker = new Worker(
-			// The URL is a pointer to a stringified function (as a blob object)
+		// The URL is a pointer to a stringified function (as a blob object)
 			URL.createObjectURL(
 				new Blob([
 					// Register our wrapper function as the message handler
 					'onmessage=(' + (
 						// userFunc() is the user-supplied async function
 						userFunc => e => {
+							// To check whether an obj is transferable
+							function isTransferable(obj) {
+								// https://developer.mozilla.org/en-US/docs/Web/API/Transferable
+								return (
+									obj instanceof ArrayBuffer ||
+									obj instanceof MessagePort ||
+									obj instanceof ImageBitmap
+								);
+							}
 							// Invoking within then() captures exceptions in userFunc() as rejections
 							Promise.resolve(e.data[1]).then(
 								userFunc.apply.bind(userFunc, userFunc)
 							).then(
 								// success handler - callback(id, SUCCESS(0), result)
-								d => { postMessage([e.data[0], 0, d]); },
+								d => {
+									if (isTransferable(d)) {
+										postMessage([e.data[0], 0, d], [d]);
+									}
+									else {
+										postMessage([e.data[0], 0, d]);
+									}
+								},
 								// error handler - callback(id, ERROR(1), error)
-								e => { postMessage([e.data[0], 1, ''+e]); }
```

Comment:
e.data on this line was giving me an error so I changed this `e` to `er` to get the event `e` from the surrounding scope.

Reply:
Good catch.